src/betterproto/__init__.py

```diff
@@ -1027,6 +1075,114 @@ def parse(self: T, data: bytes) -> T:
+            raise TypeError("`stream` must be in binary mode but isn't")
+        elif not stream.readable():
+            raise TypeError("`stream` must be readable but isn't")
+
+        encoded = stream.read(size)
+        if encoded == b"":
+            if size is None:
+                return None
+            else:
+                raise ValueError(
+                    f"Expected message of size {size} but stream was at EOF."
+                )
+        else:
+            return self.parse(encoded)
+
+    def parse_delimited_from(self: T, stream: BinaryIO) -> Optional[T]:
+        """
+        Like :code:`parse_from`, but reads the size of the message from the stream
+        first, and then the message itself based on that. Message size is encoded as a
+        varint. Returns :code:`None` if the stream is at EOF.
+
+        Parameters
+        -----------
+        stream: :class:`BufferedReader`
+            The stream to read and parse the protobuf from.
+
+        Returns
+        --------
+        :class:`Message`
+            The initialized message.
+        """
+        if "b" not in stream.mode:
+            raise TypeError("`stream` must be in binary mode but isn't")
+        elif not stream.readable():
+            raise TypeError("`stream` must be readable but isn't")
+
+        size = read_varint(stream)
+        if size is None:
+            return None
+        else:
+            return self.parse_from(stream, size)
+
+    def write_to(self, stream: BinaryIO):
+        """
+        Writes the binary encoded Protobuf message directly to the stream. This writes
+        only the message, with no delimiting or anything else.
+
+        Parameters
+        -----------
+        stream: :class:`BufferedReader`
+            The stream to write the protobuf to.
+
+        Returns
+        --------
+        :class:`Message`
+            The initialized message.
+        """
+        if "b" not in stream.mode:
+            raise TypeError("`stream` must be in binary mode but isn't")
+        elif not stream.writable():
+            raise TypeError("`stream` must be writable but isn't")
+
+        stream.write(bytes(self))
+
+    def write_delimited_to(self, stream: BinaryIO):
+        """
+        Writes the binary encoded Protobuf message directly to the stream. This writes
+        only the message, with no delimiting or anything else.
+
+        Parameters
+        -----------
+        stream: :class:`BufferedReader`
+            The stream to write the protobuf to.
+
+        Returns
+        --------
+        :class:`Message`
+            The initialized message.
+        """
+        if "b" not in stream.mode:
+            raise TypeError("`stream` must be in binary mode but isn't")
+        elif not stream.writable():
+            raise TypeError("`stream` must be writable but isn't")
+
+        encoded = bytes(self)
+        write_varint(stream, len(encoded))
+        stream.write(encoded)
+
```

Comment:
Following the nomenclature in this class, I feel these should be `from_stream`/`to_stream`:

```suggestion
    def from_stream(self: T, stream: BinaryIO) -> T:
        """
        Parse the binary encoded Protobuf from the provided stream into this message
        instance. This returns the instance itself and is therefore assignable and
        chainable.

        Parameters
        -----------
        stream: :class:`BufferedReader`
            The stream to read and parse the protobuf from.

        Returns
        --------
        :class:`Message`
            The initialized message.
        """
        size = load_varint(stream)
        buffer = stream.read(size)
        if not buffer:
            raise EOFError(f"Reading {size} message bytes")
        return self.parse(buffer)

    def to_stream(self, stream: BinaryIO):
        """
        Writes the binary encoded Protobuf message directly to a stream. 

        Parameters
        -----------
        stream: :class:`BufferedReader`
            The stream to write the protobuf to.

        Returns
        --------
        :class:`Message`
            The initialized message.
        """
        buffer = bytes(self)
        dump_varint(len(buffer), stream)
        stream.write(buffer)
```

You may want to have:
  * `to_stream`/`from_stream`: does no delimiting
  * `to_coded_stream`/`from_coded_stream`: does delimiting
I can't think (other than testing) why you would want to not have delimiting. You _need_ the size to marshall/unmarshall the message. The only reason I can think why you would want zero delimiting is if you were storing the same message over and over again. In that case you can just do `stream.write(bytes(message))` which makes `to_stream`/`from_stream` redundant. I think making `to_stream`/`from_stream` delimiting (as in the above code snippet) makes sense.

Reply:
Good point about the naming, I'll change that now. The reason for the function that doesn't do delimiting is that that is also present in [the Java version](https://protobuf.dev/reference/java/api-docs/com/google/protobuf/MessageLite.html#writeTo-com.google.protobuf.CodedOutputStream-) so I wanted to maintain proper compatibility, plus it allows a dev flexibility on how they deal with messages.

Reply:
I've made the changes to the naming, there are now `to_stream` and `to_stream_delimited`, with the former also explicitly stating in its docstring that it doesn't do delimiting. On the removal of the non-delimited one, I haven't yet done that, but wanted to know what you thought with it being consistent with Java's equivalents like I mentioned before I changed that.

Reply:
> The reason for the function that doesn't do delimiting is that that is also present in [the Java version](https://protobuf.dev/reference/java/api-docs/com/google/protobuf/MessageLite.html#writeTo-com.google.protobuf.CodedOutputStream-) so I wanted to maintain proper compatibility, plus it allows a dev flexibility on how they deal with messages.

I would say that we likely don't want to completely match the Java version for few reasons:

* The project _already_ doesn't use the same nomenclature as Java/C++
    * However, it _does_ have some compatibilty functions (`ParseFromString`, etc) that match the [Python version](https://googleapis.dev/python/protobuf/latest/google/protobuf/message.html)
    * The Python version does not have a coded stream equivalent.
* `writeToDelimited` is Java specific, it does not occur in [C++](https://protobuf.dev/reference/cpp/api-docs/google.protobuf.message/#Message) message

Hence, I feel that we should just do what we think is most correct here. I'd like to posit that the `to_stream`/`from_stream` seem redundant to add to the API of `Message`. 

The `to_stream`/`from_stream` are extremely trivial to implement as a user around the `Message` class:

```
buffer = bytes(message)
size = len(buffer)
stream.write(buffer)
buffer = stream.read(size)
message = Abc.parse(buffer)
```

This is an extremely constrained case. Due to every serialisation of a message being different sizes, you cannot reliably marshall/unmarshall multiple messages of the same type:

```
message.string = "something"
buffer = bytes(message)
size = len(buffer)
stream.write(buffer)
message.string = "something else"
buffer = bytes(message)
assert size == len(buffer)  # Nope, this will be 5 bytes longer to encode the ` else`
```

So writing these two messages to the stream would be impossible to unmarhsall on the other end because you cannot know what size to read. This makes the `from_stream(size)` hard to use in the general case without delimiting so it seems redundant.

My thinking is that you pretty much _always_ want delimiting. We want the `to_stream`/`from_stream` to be safe. Hence, we should implement `from_stream`/`to_stream` as the delimited versions. If a user wants to do custom marshalling/unmarshalling, they can use the `load`/`dump` variable integer functions and `bytes(message)`. If we need single message serialisation/deserialisation (serde) in the tests we can use `bytes(message)` and `read` with a fixed, expected, length.

I would like us to make APIs are clear, concise and easy to use. `from_stream`/`to_stream` are trigging that right now as I'm struggling to see the use for them when there are clear, simple, alternatives and that the usage of them doesn't have a general case.